x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -941,72 +1005,49 @@ public void onFailure(Exception e) {
+            final String supersedingAccessToken = new String(cipher.doFinal(encryptedSupersedingAccessToken), StandardCharsets.UTF_8);
+            byte[] iv2 = new byte[iv.length];
+            System.arraycopy(iv, iv.length / 2, iv2, 0, iv.length / 2);
+            System.arraycopy(iv, 0, iv2, iv.length / 2, iv.length / 2);
+            cipher = getDecryptionCipher(iv2, refreshToken, salt);
+            final String supersedingRefreshToken = new String(cipher.doFinal(encryptedSupersedingRefreshToken), StandardCharsets.UTF_8);
+            listener.onResponse(new Tuple<>(prependVersionAndEncodeAccessToken(refreshTokenStatus.getVersion(), supersedingAccessToken),
+                prependVersionAndEncodeRefreshToken(refreshTokenStatus.getVersion(), supersedingRefreshToken)));
+        } catch (GeneralSecurityException | IOException e) {
+            logger.warn("Could not get stored superseding token values", e);
+            listener.onFailure(invalidGrantException("could not refresh the requested token"));
+        }
     }
 
-    private void getSupersedingTokenDocAsync(RefreshTokenStatus refreshTokenStatus, ActionListener<GetResponse> listener) {
-        final String supersedingDocReference = refreshTokenStatus.getSupersededBy();
-        if (supersedingDocReference.startsWith(securityTokensIndex.aliasName() + "|")) {
-            // superseding token doc is stored on the new tokens index, irrespective of where the superseded token doc resides
-            final String supersedingDocId = supersedingDocReference.substring(securityTokensIndex.aliasName().length() + 1);
-            getTokenDocAsync(supersedingDocId, securityTokensIndex, listener);
-        } else {
-            assert false == supersedingDocReference
-                    .contains("|") : "The superseding doc reference appears to contain an alias name but should not";
-            getTokenDocAsync(supersedingDocReference, securityMainIndex, listener);
-        }
+    /*
+     * Encrypts the values of the superseding access token and the refresh token, using a key derived from the superseded refresh token.
+     */
+    Tuple<String, String> encryptSupersedingTokens(String supersedingAccessToken, String supersedingRefreshToken,
+                                                   String refreshToken, byte[] iv, byte[] salt) throws GeneralSecurityException {
+        Cipher cipher = getEncryptionCipher(iv, refreshToken, salt);
+        final String encryptedSupersedingAccessToken =
+            Base64.getEncoder().encodeToString(cipher.doFinal(supersedingAccessToken.getBytes(StandardCharsets.UTF_8)));
+        // In AES GCM we cannot reuse the same IV. We predictably generate the IV for the second decryption instead of
+        // storing an extra field, since it doesn't have to be unpredictable, just not reused with the same key.
+        byte[] iv2 = new byte[iv.length];
+        System.arraycopy(iv, iv.length / 2, iv2, 0, iv.length / 2);
```

Comment:
This has no particular meaning, we just need a predictable way to get a new IV for the second encryption,